### DISCUSSION TITLE
Rootless docker with account in sudoers group and passwordless sudo

### DISCUSSION THREAD

**Discussion by lonix1** [moby/moby#45632] [Q&A]
I want to use use docker in ["rootless mode"](https://docs.docker.com/engine/security/rootless). As I understand it, if a container (or the daemon itself) is compromised, then the code that breaks out into the host can only do damage limited to the access rights of the user running the docker daemon (my user account).

But on a VPS (especially one controlled by automation, e.g. ansible), it is common to have the user account in the `sudo` group, and for it to have passwordless sudo (as specified in `/etc/sudoers`).

Are these contradictory concepts? Is rootless mode still useful, given that malware breaking out into the host (and thus my user account) will have passwordless sudo rights?

**Comment by cpuguy83** [comment]
If you have passwordless sudo then there is nothing preventing something from escalating permissions.
BUT...

The way rootless works is by running dockerd inside of a user namespace where root inside that user namespace is mapped to the UID of the user who started dockerd. So assume that any breakout from the root user in that namespace will escalate to whatever permissions you have on the host.
You could, in theory i haven't tested this, use a non-root user inside your containers and those are (should be? It's been awhile since I looked at this) mapped to entirely different UID's on the host and as such it becomes much harder to escalate and use passwordless sudo.

NOW, all that being said, all of this depends on user namespaces... and in particular a kernel feature that allows an unprivileged user to create a new user namespace.
This code path in the kernel is not well exercised and indeed there have been multiple root escalations (even one in the last couple of weeks) vulnerabilities caused by this feature.
So take any security claims for rootless with a grain of salt.

**Reply by AkihiroSuda** [reply]
> This code path in the kernel is not well exercised and indeed there have been multiple root escalations (even one in the last couple of weeks) vulnerabilities caused by this feature.
> So take any security claims for rootless with a grain of salt.

Yes, but it is still much secure than running everything as root.

Anyway, I'm trying to implement a fine-grained control knob for unprivileged user namespaces: [`[PATCH] userns: add sysctl "kernel.userns_group_range"`](https://lore.kernel.org/lkml/168547265011.24337.4306067683997517082-0@git.sr.ht/).
With this kernel patch, unprivileged UserNS can be enabled only for user accounts who need containers, and can be disabled for other accounts on the same host.
